### ISSUE TITLE
Complete post-mortem of dcp12 reindex

### ISSUE THREAD

**Issue by danielsotirhos** [DataBiosphere/azul#3703]
A reindex of the `dcp12` catalog was performed Thursday 12/16 and retries on Friday 12/17 as part of the prod promotion https://github.com/DataBiosphere/azul/pull/3697

- 12/15 3:30p : Prod promotion performed

- 12/16 6:30a : Reindex of dcp12 catalog
Result: azul-notifications_fail-prod    441 entries

- 12/17 4:00a : Reindex of 441 failures
Result: azul-notifications_fail-prod    382 entries

- 12/17 8:00a : Reindex of 382 failures
Result: azul-notifications_fail-prod    377 entries

[azul-notifications_fail-prod.377.json.txt](https://github.com/DataBiosphere/azul/files/7738433/azul-notifications_fail-prod.377.json.txt)

The CloudWatch Metrics (`Throttles` in particular) for this reindex of `dcp12`...

![Screen Shot 2021-12-17 at 4 42 49 PM](https://user-images.githubusercontent.com/46407464/146623171-f8c6db02-75f1-47d2-a45b-adc15870123c.png)

...have a much different shape than the previous reindex (of `dcp11`) on 12/10...

![Screen Shot 2021-12-17 at 4 41 41 PM](https://user-images.githubusercontent.com/46407464/146623397-392c6ffb-64bb-440a-ad9b-f36947074892.png)

...and the reindexes prior to that on 12/01 and 12/03.

![Screen Shot 2021-12-17 at 4 47 06 PM](https://user-images.githubusercontent.com/46407464/146623434-8928b881-406a-49c5-8764-a1ab89333a87.png)

**Comment by hannes-ucsc** [comment]
We identified the following issues:

1) Indexing the dcp12 snapshot requires three times more slots to get the same throughput from BQ as the last reindex of the dcp11 snapshot. The first dcp12 reindex done by @danielsotirhos showed a 2.5 slowdown. A repeated attempt to reindex initially showed the same slowdown but throughput increased after to expected levels after buying 600 more slots (900 total)

2) Refeeding several hundred failed notifications caused a high degree of throttles. A throttle occurs when AWS' internal SQS-Lambda integration machinery fetches more messages than can be handled with a given concurrency limit (64 in this case). The docs state that the machinery has 5 threads pulling from the queue in parallel. There is evidence for that in the "Empty receives" metric for SQS: we observe 15 = 5 * 60 / 20 empty receives per queue and minute if the system is idle. Each thread produces 3 empty receives per minute because it uses 20s long polling. If there are more than 64 notifications in the retry queue, the machinery will receive all messages but it will only be able to allocate Lambda executions for 64 of them. The rest will be returned to the queue with an incremented receive count that doesn't reflect the number of actual attempts to process the message. This is an unfortunate design choice by AWS. They could easily return the message without incrementing the receive count. Since the receive count is capped at 9 for the retry queues, we don't get an honest 9 attempts. We observed messages to end up in the fail queue with only four honest attempts.

3) After the second reindex of the dcp12, there were some genuine timeouts that were fixed by just retrying them:
[azul-notifications_fail-prod.not_in_5b.json.txt](https://github.com/DataBiosphere/azul/files/7759128/azul-notifications_fail-prod.not_in_5b.json.txt)

4) The remaining failed notifications all belonged to project `5b328561-4a97-40ac-b7ad-6a90fc59d374`. They timed out because of https://github.com/DataBiosphere/azul/issues/3705 and a hotfix on prod fixed those.
